### PR TITLE
Implement an Events interface for the default class

### PR DESCRIPTION
**Pull request by smellyshovel**
This adds support for listening to maplibre gl directions events.

Currently implemented events are: 

-   setwaypoints
-   addwaypoint
-   removewaypoint
-   movewaypoint
-   fetchroutesstart
-   fetchroutesend

This is a bare minimum and the list will probably be extended in the future, but these a the minimal required amount to unblock other features (such as controls).

### CODE REVIEW

File: demo/src/examples/7 Events.vue [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    map.on("load", () => {
+      directions.value = new MaplibreGlDirections(map, {
+        requestOptions: {
+          alternatives: "true",
+        },
+      });
+
+      directions.value.interactive = true;
+
+      directions.value?.on("addwaypoint", (e) => {
```
Comment (wipfli) [inline demo/src/examples/7 Events.vue:55]: What does the question mark do?

Comment (smellyshovel) [reply demo/src/examples/7 Events.vue:55]: Good point :+1:, it's an artifact of copying the code around. Thanks for noticing!

**Comment by smellyshovel** [comment]
@wipfli if we're good with this one, let's approve and merge it and afterwards check the new automated release workflow out. What do you think?

**Comment by smellyshovel** [comment]
Removed the "semver:minor" and added the "semver:patch" because when major is less than 1, minor changes are reserved for backwards-incompatible changes. And this  one is a new feature (backwards-compatible).

File: src/directions/main.ts [hunk 2]
```diff
@@ -672,42 +695,51 @@ export default class MaplibreGlDirections {
...
    *
-   * @param {[number, number]} waypoint The coordinates at which the waypoint should be added
-   * @param {number?} index The index the waypoint should be inserted at. If omitted, the waypoint is inserted at the end
-   * @return {Promise<void>} Resolved after the routing request has finished
+   * @param waypoint The coordinates at which the waypoint should be added
+   * @param index The index the waypoint should be inserted at. If omitted, the waypoint is inserted at the end
+   * @return Resolved after the routing request has finished
    */
-  async addWaypoint(waypoint: [number, number], index?: number) {
+  async addWaypoint(waypoint: [number, number], index?: number, originalEvent?: MapMouseEvent | MapTouchEvent) {
```
Comment (wipfli) [inline src/directions/main.ts:712]: Document originalEvent in jsdoc

File: src/directions/main.ts [hunk 3]
```diff
@@ -672,42 +695,51 @@ export default class MaplibreGlDirections {
...
   /**
    * Removes a waypoint and its related snappoint by the waypoint's index from the map and re-fetches the routes.
    *
-   * @param {number} index The index of the waypoint to remove
-   * @return {Promise<void>} Resolved after the routing request has finished
+   * @param index The index of the waypoint to remove
+   * @return Resolved after the routing request has finished
    */
-  async removeWaypoint(index: number) {
+  async removeWaypoint(index: number, originalEvent?: MapMouseEvent | MapTouchEvent) {
```
Comment (wipfli) [inline src/directions/main.ts:732]: document originalEvent

**Review by wipfli — APPROVED** [review]
Sounds like a good idea with the patch release. Please document the originalEvent and then this is good from my side.

File: src/directions/main.ts (hunk 2 above)
Comment (smellyshovel) [reply src/directions/main.ts:712]: @wipfli it left un-documented specifically. The end user shouldn't be using this parameter, it's only for internal-purposes. I thought of adding another wrapper-method with that third parameter, but I think it'd be an overkill.

File: src/directions/main.ts (hunk 3 above)
Comment (smellyshovel) [reply src/directions/main.ts:732]: Same as above...

File: src/directions/main.ts (hunk 2 above)
Comment (wipfli) [reply src/directions/main.ts:712]: Interesting thanks. In that case, please add a comment.

Comment (wipfli) [reply src/directions/main.ts:712]: Something like "for internal use only"

**Comment by smellyshovel** [comment]
@wipfli could you please double-check one more time? In the end I decided to move the functionality of `addWaypoint` and `removeWaypoint` into respective protected methods (which have this 3rd parameter), and the public methods simply invoke the protected ones without the last param. That way we don't need to make any docs-related hacks to hide the docs for the intrinsic-params.

**Review by wipfli — APPROVED** [review]

**Comment by wipfli** [comment]
Then let's see if the release automation works :)

**Pull request merged**